File: chapter_1_arrays_strings/c/1.2-check_permutation.c
#include <stdio.h>
#include <string.h>

#define ASCII_LENGTH 128

/* O(n) solution */
int check_permutation(char *str1, char *str2)
{
    int len1, len2, i;
    int map[ASCII_LENGTH] = {0};

    len1 = strlen(str1);
    len2 = strlen(str2);

    if (len1 != len2)
        return 0;
    for (i = 0; i < len1; i++)
        map[str1[i]] += 1;
    for (i = 0; i < len2; i++)
    {
        map[str2[i]]--;
        if (map[str2[i]] < 0)
            return 0;
    }
    return 1;
}

int main()
{
    printf("%i\n", check_permutation("dog", "god"));     // 1
    printf("%i\n", check_permutation("dog", "God"));     // 0
    printf("%i\n", check_permutation("dog", "good"));    // 0
    printf("%i\n", check_permutation("mmmmm", "mmmmm")); // 1
    printf("%i\n", check_permutation("mmmmm", "mmmmn")); // 0

    return 0;
}

File: chapter_1_arrays_strings/python/1.2-check_permutation.py
# O(n) solution


def create_map(string):
    obj = {}
    for ch in string:
        if ch in obj:
            obj[ch] += 1
        else:
            obj[ch] = 1
    return obj


def check_permutation(str1, str2):
    if len(str1) != len(str2):
        return False
    str_map = create_map(str1)
    for ch in str2:
        if ch not in str_map:
            return False
        str_map[ch] -= 1
        if str_map[ch] < 0:
            return False
    return True


print(check_permutation('dog', 'god'))  # true
print(check_permutation('dog', 'God'))  # false
print(check_permutation('dog', 'good'))  # false
print(check_permutation('mmmmm', 'mmmmm'))  # true
print(check_permutation('mmmmm', 'mmmmn'))  # true
